File: 总题库/407.TrappingRainWater2.cpp
struct Node{
    int i, j, h;
    Node(int x, int y, int z):i(x), j(y), h(z){}
    bool operator < (const Node& b) const {
        if (h > b.h) return true;
        return false;
    }
};

class Solution {
public:
    int trapRainWater(vector<vector<int>>& heightMap) {
        if (heightMap.size() == 0 || heightMap[0].size() == 0) {
            return 0;
        }
        int rows = heightMap.size(), cols = heightMap[0].size();
        vector<vector<bool>> visited(rows, vector<bool>(cols, false)); // visited数组，用来标记某个节点是否被访问；
        priority_queue<Node> pq; // 优先队列，用来维护围墙边界；
        int ans = 0; // 用来储存结果；

        // 首先存入最外层边界，然后依此往里面缩小；
        for (int i = 0; i < rows; ++i) {
            pq.push({i, 0, heightMap[i][0]});
            pq.push({i, cols - 1, heightMap[i][cols - 1]});
            visited[i][0] = true;
            visited[i][cols - 1] = true;
        }
        for (int i = 0; i < cols; ++i) {
            pq.push({0, i, heightMap[0][i]});
            pq.push({rows - 1, i, heightMap[rows - 1][i]});
            visited[0][i] = true;
            visited[rows - 1][i] = true;
        }
        
        int dirs[5] = {-1, 0, 1, 0, -1};
        while (!pq.empty()) {
            Node cur = pq.top();
            pq.pop();
            for (int i = 0; i < 4; ++i) {
                int newX = cur.i + dirs[i], newY = cur.j + dirs[i + 1];
                if (newX >= 0 && newX < rows && newY >= 0 && newY < cols && visited[newX][newY] == false) {
                    if (cur.h > heightMap[newX][newY]) {
                        // 围墙高度大于相邻围墙的高度，可以灌水；
                        ans += cur.h - heightMap[newX][newY];
                        heightMap[newX][newY] = cur.h;
                    }
                    pq.push({newX, newY, heightMap[newX][newY]});
                    visited[newX][newY] = true;
                }
            }
        }
        return ans;
    }
};

File: 总题库/407.TrappingRainWater2.py
import heapq # Python3中的堆操作；

class Solution:
    def trapRainWater(self, heightMap: List[List[int]]) -> int:
        if not len(heightMap) or not len(heightMap[0]):
            return 0

        rows, cols = len(heightMap), len(heightMap[0]);
        pq = [] # pq is a priority queue, implemented with heap.
        visited = [[False for _ in range(cols)] for _ in range(rows)]
        dirs = [-1, 0, 1, 0, -1]
        ans = 0

        # 首先存入最外圈的边界
        for i in range(rows):
            heapq.heappush(pq, (heightMap[i][0], i, 0))
            heapq.heappush(pq, (heightMap[i][cols - 1], i, cols - 1))
            visited[i][0] = True
            visited[i][cols - 1] = True

        for i in range(cols):
            heapq.heappush(pq, (heightMap[0][i], 0, i))
            heapq.heappush(pq, (heightMap[rows - 1][i], rows - 1, i))
            visited[0][i] = True
            visited[rows - 1][i] = True

        while pq:
            cur = heapq.heappop(pq)

            for i in range(4):
                newX, newY = cur[1] + dirs[i], cur[2] + dirs[i + 1]
                if newX >= 0 and newX < rows and newY >= 0 and newY < cols and visited[newX][newY] == False:
                    if cur[0] > heightMap[newX][newY]:
                        ans += cur[0] - heightMap[newX][newY]
                        heightMap[newX][newY] = cur[0]
                    heapq.heappush(pq, (heightMap[newX][newY], newX, newY))
                    visited[newX][newY] = True

        return ans